fabric-tool-attribute-api-v1/src/testmod/java/net/fabricmc/fabric/test/tool/attribute/ToolAttributeTest.java

```diff
@@ -18,63 +18,122 @@
+	private static final float TOOL_BREAK_SPEED = 10.0F;
+
+	private boolean hasValidated = false;
+
+	Block gravel_block;
+	Block stone_block;
+	Item test_shovel;
+	Item test_pickaxe;
 
 	@Override
 	public void onInitialize() {
 		// Register a custom shovel that has a mining level of 2 (iron) dynamically.
-		Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "test_shovel"), new TestShovel(new Item.Settings()));
+		test_shovel = Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "test_shovel"), new TestTool(new Item.Settings(), FabricToolTags.SHOVELS));
+		//Register a custom pickaxe that has a mining level of 2 (iron) dynamically.
+		test_pickaxe = Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "test_pickaxe"), new TestTool(new Item.Settings(), FabricToolTags.PICKAXES));
 		// Register a block that requires a shovel that is as strong or stronger than an iron one.
-		Block block = Registry.register(Registry.BLOCK, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_block"),
+		gravel_block = Registry.register(Registry.BLOCK, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_gravel_block"),
 				new Block(FabricBlockSettings.of(new FabricMaterialBuilder(MaterialColor.SAND).build(), MaterialColor.STONE)
 						.breakByTool(FabricToolTags.SHOVELS, 2)
 						.requiresTool()
 						.strength(0.6F)
 						.sounds(BlockSoundGroup.GRAVEL)));
-		Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_block"), new BlockItem(block, new Item.Settings()));
+		Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_gravel_block"), new BlockItem(gravel_block, new Item.Settings()));
+		// Register a block that requires a pickaxe that is as strong or stronger than an iron one.
+		stone_block = Registry.register(Registry.BLOCK, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_stone_block"),
+				new Block(FabricBlockSettings.of(Material.STONE, MaterialColor.STONE)
+						.breakByTool(FabricToolTags.PICKAXES, 2)
+						.requiresTool()
+						.strength(0.6F)
+						.sounds(BlockSoundGroup.STONE)));
+		Registry.register(Registry.ITEM, new Identifier("fabric-tool-attribute-api-v1-testmod", "hardened_stone_block"), new BlockItem(stone_block, new Item.Settings()));
 
-		ServerTickCallback.EVENT.register(this::validateTags);
+		ServerTickCallback.EVENT.register(this::validate);
 	}
 
-	private void validateTags(MinecraftServer server) {
-		if (hasValidatedTags) {
+	private void validate(MinecraftServer server) {
+		if (hasValidated) {
 			return;
 		}
 
-		hasValidatedTags = true;
+		hasValidated = true;
 
 		if (FabricToolTags.PICKAXES.values().isEmpty()) {
 			throw new AssertionError("Failed to load tool tags");
 		}
+
+		//Test vanilla tools don't bypass fabric mining levels
```

Comment:
possibly want to test for vanilla tools for vanilla blocks

Reply:
Good idea, I've added all vanilla test cases.